lib/client.js

```diff
@@ -263,7 +263,23 @@ Client.prototype._buildForward = function() {
       }
 
       if (qos && options.messageId) {
-        that.server.updateOfflinePacket(that, options.messageId, packet, doForward);
+
+        //converting the payload to a JSON object
+        var packetPayload = JSON.parse(packet.payload.toString());
```

Comment:
@behrad my bad, updated the code handle the cases where the payload might not be a valid JSON.
@mcollina This change does not force the clients to send the messageId inside the payload, this is limited to the internal working of mosca. Clients may continue publishing messages as they wish.

This messageId will be used internally by mosca only for updating the persistent storage correctly. Once the message has been received this messageIs is deleted from the payload and then forward to the actual client subscribing to the topic